test/e2e/gateway/008_tlsroute_test.go

```diff
@@ -100,3 +100,71 @@ func testTLSRoutePassthrough(fx *e2e.Framework, namespace string) {
 	assert.Truef(t, ok, "expected 200 response code, got %d", res.StatusCode)
 	assert.Equal(t, "echo", fx.GetEchoResponseBody(res.Body).Service)
 }
+
+func testTLSRouteTerminate(fx *e2e.Framework, namespace string) {
+	t := fx.T()
+
+	fx.CreateNamespace(namespace)
+	defer fx.DeleteNamespace(namespace)
+
+	fx.Fixtures.Echo.Deploy(namespace, "echo")
+
+	route := &gatewayv1alpha1.TLSRoute{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: namespace,
+			Name:      "tls-route-1",
+		},
+		Spec: gatewayv1alpha1.TLSRouteSpec{
+			Gateways: &gatewayv1alpha1.RouteGateways{
+				Allow: gatewayAllowTypePtr(gatewayv1alpha1.GatewayAllowAll),
+			},
+			Rules: []gatewayv1alpha1.TLSRouteRule{{
+				Matches: []gatewayv1alpha1.TLSRouteMatch{{
+					SNIs: []gatewayv1alpha1.Hostname{
+						gatewayv1alpha1.Hostname("tlsroute.gatewayapi.projectcontour.io"),
+					},
+				}},
+				ForwardTo: []gatewayv1alpha1.RouteForwardTo{{
+					ServiceName: stringPtr("echo"),
+					Port:        portNumPtr(80),
+				}},
+			}},
+		},
+	}
+	fx.CreateTLSRouteAndWaitFor(route, tlsRouteAdmitted)
+
+	// Ensure request routes to echo.
+	res, ok := fx.HTTP.SecureRequestUntil(&e2e.HTTPSRequestOpts{
+		Host:      "tlsroute.gatewayapi.projectcontour.io",
+		Condition: e2e.HasStatusCode(200),
+	})
+	assert.Truef(t, ok, "expected 200 response code, got %d", res.StatusCode)
+	assert.Equal(t, "echo", fx.GetEchoResponseBody(res.Body).Service)
+
+	require.NoError(t, retry.RetryOnConflict(retry.DefaultBackoff, func() error {
```

Comment:
a quick comment indicating what's happening here would be helpful (looks like the SNI match is being dropped and the routing is expected to still work).

Reply:
I wonder if it's still working or the SNI is added automatically. I can update with the SNI bit (now that I see it's in the test suite) since that's the expected behavior.

Reply:
ok updated with SNI

Reply:
Sorry, I was talking about https://github.com/projectcontour/contour/blob/f53e69113238ab1a6714a7ba507f575ee66f1fca/test/e2e/gateway/008_tlsroute_test.go#L144-L161 (GH doesn't always make it clear what line you commented on) -- that updates the TLSRouteRule to *not* have the `Matches` block (https://github.com/projectcontour/contour/blob/f53e69113238ab1a6714a7ba507f575ee66f1fca/test/e2e/gateway/008_tlsroute_test.go#L122-L126), only to have the `ForwardTo`. So seems like that's verifying that if you don't explicitly specify a match on SNI, then any SNI will still route traffic there.

Reply:
Ahh ok I'm following now. I updated some comments. Also added a negative test to validate a non-matching SNI doesn't produce a response.

Reply:
:+1: nice